File: lsystem/LSystem.py
import math
from turtle.Turtle import *
from Conf import Conf
import pygame
from OpenGL.GL import shaders

from OpenGL.GL import *
from OpenGL.GLU import *
from OpenGL.GL.ARB.framebuffer_object import *
from OpenGL.GL.EXT.framebuffer_object import *
from OpenGL.GL.ARB.vertex_buffer_object import *
from OpenGL.GL.ARB.geometry_shader4 import *
from OpenGL.GL.EXT.geometry_shader4 import *

class LSystem(object):
	"""Abstract class: allow generation of a L-System
	but does not own rules or variables
	You need to inherit from this class to implement your own l-system"""

	def __init__(self, turtle):
		self.turtle = turtle
		self.LSName = "Undefined" 	# nom de la fractale
		self.LSRules = {}		  	# regles de generation
		self.LSVars = {}		  	# variables = pointeurs sur actions de la turtle
		self.LSParams = {}		  	# parametres a appliquer aux actions de la turtle
		self.LSAngle = math.pi / 2	# angle (pour les rotations)
		self.LSSegment = 1.0		# taille d'un segment
		self.LSSteps = 5			# nombre de passe de generation (= precision de la fractale)
		self.LSStartingString = "F"	# axiome initial
		self.LSCode = ""			# code genere
		self.LSStochastic = False	# determine si le dessin doit etre aleatoire
		self.LSStochRange = 0.01	# determine le facteur aleatoire
		self.LSVertexShader = ""	# vertex shader a appliquer a la fractale
		self.LSPixelShader = ""		# pixel shader a appliquer a la fractale
		self.LSUniforms = {}		# association nom-fonction des donnees uniform a associer au shader de cette fractale
		self.LSDrawType = GL_LINE_STRIP	# la facon dont les vertices du shader seront dessinees
		self.shaderProgram = None
		self.defineParams()
		self.createVars()
		self.createRules()
		self.currentMaxStep = 0
		self.currentStep = 0
		self.autorun = False
		self.stepbystep = False
		self.createShaders()
		self.compileShaders()

	""" This method should set the following attributes of the fractal:
			- LSName: this is the name of the fractal
			- LSAngle: this is the default angle applied to rotations
			- LSSegment: this is the default segment length applied to forward or backward steps
			- LSSteps: this is the default number of steps done when recursively generating the code
			- LSStartingString: this is the starting code
			- LSDrawType: the way the fractal is drawed (GL_LINES, GL_LINE_STRIP, GL_TRIANGLES, GL_TRIANGLE_STRIP, ...)
	 """
	def defineParams(self):
		raise NotImplementedError("The method `defineParams' of the LSystem `" + self.LSName + "' has not been implemented yet")

	""" This method should fill in the LSVars dico that will associate actions of the turtle to some variables
	 and the LSParams dico what argument each function of the LSVars dico need  """
	def createVars(self):
		raise NotImplementedError("The method `createVars' of the LSystem `" + self.LSName + "' has not been implemented yet")

	""" This method should fill in the LSRules dico that will be used to recursively generate the fractal code"""
	def createRules(self):
		raise NotImplementedError("The method `createRules' of the LSystem `" + self.LSName + "' has not been implemented yet")

	""" This method can change the default shaders applied to the fractal, and eventually set the LSUniforms associative array """
	def createShaders(self):
		self.LSVertexShader = """
// varying vec4 gl_Vertex;
// varying vec4 gl_Color;
uniform float time; // exemple of giving a uniform value
varying vec4 vertex_color;
void main() {
	gl_Position = gl_ModelViewProjectionMatrix * gl_Vertex;
	vertex_color = gl_Color;
}
	    """
		self.LSPixelShader = """
varying vec4 vertex_color;
void main() {
	gl_FragColor = vertex_color;
}
	    """
		self.LSUniforms = {
	    	'time': lambda: time.time() # a lambda function is used so that each time the shader is rendered, the value change
    	}

	#compile shaders
	# important note : if this function is not called after an openGL context has been defined,
	# it could crash the program with an error telling that a function is NULL
	def compileShaders(self):
		if not 'vec3 rgb2hsv(vec3 c)' in self.LSVertexShader:
			self.LSVertexShader = """
vec3 rgb2hsv(vec3 c)
{
	vec4 K = vec4(0.0, -1.0 / 3.0, 2.0 / 3.0, -1.0);
	vec4 p = mix(vec4(c.bg, K.wz), vec4(c.gb, K.xy), step(c.b, c.g));
	vec4 q = mix(vec4(p.xyw, c.r), vec4(c.r, p.yzx), step(p.x, c.r));

	float d = q.x - min(q.w, q.y);
	float e = 1.0e-10;
	return vec3(abs(q.z + (q.w - q.y) / (6.0 * d + e)), d / (q.x + e), q.x);
}

vec3 hsv2rgb(vec3 c)
{
	vec4 K = vec4(1.0, 2.0 / 3.0, 1.0 / 3.0, 3.0);
	vec3 p = abs(fract(c.xxx + K.xyz) * 6.0 - K.www);
	return c.z * mix(K.xxx, clamp(p - K.xxx, 0.0, 1.0), c.y);
}
""" + self.LSVertexShader
		# if not '#version' in self.LSVertexShader:
		# 	self.LSVertexShader = '#version 320\n' + self.LSVertexShader
		# if not '#version' in self.LSPixelShader:
		# 	self.LSPixelShader = '#version 320\n' + self.LSPixelShader
		# VERTEX_SHADER = shaders.compileShader(self.LSVertexShader, GL_VERTEX_SHADER)

		# FRAGMENT_SHADER = shaders.compileShader(self.LSPixelShader, GL_FRAGMENT_SHADER)

		# self.shader = shaders.compileProgram(VERTEX_SHADER, FRAGMENT_SHADER)

		vs = glCreateShader(GL_VERTEX_SHADER)
		fs = glCreateShader(GL_FRAGMENT_SHADER)
		#gs = glCreateShader(GL_GEOMETRY_SHADER_EXT)

		glShaderSource(vs, self.LSVertexShader)
		glShaderSource(fs, self.LSPixelShader)
		# glShaderSource(gs, source['geometry'])

		glCompileShader(vs)
		log = glGetShaderInfoLog(vs)
		if log:
			print "Vertex Shader: "
			print self.LSVertexShader
			print '>>>> Compilation Logs <<<< '
			print log

		# glCompileShader(gs)
		# log = glGetShaderInfoLog(gs)
		# if log: print 'Geometry Shader: ', log

		glCompileShader(fs)
		log = glGetShaderInfoLog(fs)
		if log:
			print "Pixel Shader: "
			print self.LSPixelShader
			print '>>>> Compilation Logs <<<< '
			print log

		self.shader = glCreateProgram()

		glAttachShader(self.shader, vs)
		glAttachShader(self.shader, fs)
		# glAttachShader(self.shader, gs)

		glLinkProgram(self.shader)

	def getShader(self):
		return self.shader

	def getUniforms(self):
		return self.LSUniforms

	def getCode(self):
		return self.LSCode

	def setSteps(self, n):
		if n < 1:
			self.LSSteps += n
		if self.LSSteps < 1:
			self.LSSteps = 1
		if n >= 1:
			self.LSSteps = n

	""" This method will recursively generate the code of the fractal"""
	def generate(self):
		print "============= Fractal Name: ", self.LSName
		print "============= Fractal Precision: ", self.LSSteps
		if self.LSStochastic:
			print "============= Fractal Stochastic Factor: ", self.LSStochRange
		print "============= LSystem Inital Axiom: ", self.LSStartingString
		print "============= LSystem Rules: "
		for r in self.LSRules:
			print '\t', r, "=>", self.LSRules[r]
		print "============= Starting generation. "
		self.LSCode = self.LSStartingString
		self.generate_param_rec(self.LSSteps)
		if self.LSStochastic:
			self.turtle.setStochasticFactor(self.LSStochRange)
		self.LSCodeLen = len(self.LSCode)
		print "============= Generation DONE !"

	""" Run the turtle over the code of the fractal """
	def runTurtleRun(self, stepbystep=False):
		print "Starting turtle !"
		self.stepbystep = stepbystep

		self.turtle.setDrawType(self.LSDrawType)

		# initialize step by step engine
		if self.stepbystep:
			self.currentStep = 0
			self.currentMaxStep = 0
			self.inc_max_step()

		# Begin the command sequence
		self.turtle.begin()
		if self.stepbystep:
			while self.currentStep < self.currentMaxStep:
				tmp = [self.currentStep];
				self.execute(tmp)
				self.currentStep = tmp[0]
		else:
			count = [0]
			while count[0] < self.LSCodeLen:
				#count[0] += 1
				if count[0] % 50000 is 0:
					print "Turtle is running: ",
					percent = (float(count[0]) / float(self.LSCodeLen) * 100)
					print "%.2f" % percent,
					print "% => [",
					print self.turtle.vertexBufferLength, "] vertex"
				self.execute(count)
		self.turtle.end() # end the commend sequence

	# execute the current instruction, using the eventual parameters
	# pos will be incremented
	def execute(self, pos):
		dbg = pos[0]
		char = self.LSCode[pos[0]];
		if pos[0] + 1 < self.LSCodeLen and self.LSCode[pos[0] + 1] == '(':
			#print "Step: ", pos[0], ": ",
			arg = self.find_arg(pos) # extract arguments of the instruction
			if arg is None:
				print "Error: no argument found for inst `"+char+"' in: ", self.LSCode[dbg:dbg+10]
				return;
			# if we are executing a parametized rule:
			# replace each variable by its value in the params array
			for param in self.LSParams:
				if param in arg:
					arg = arg.replace(param, str(float(self.LSParams[param])))
			res = eval(arg) # evaluate expression

			#print self.LSVars[char].__name__ + "(" + str(res) + ")"

			if char in self.LSVars:
				self.LSVars[char](res)	# call corresponding function, giving the parameter
		else:
			#print "Step: ", pos[0], ": ",
			#print self.LSVars[char].__name__ + "(" + str(self.LSParams[char]) + ")"
			if char in self.LSVars:
				self.LSVars[char](self.LSParams[char])	# call corresponding function, giving the parameter
			pos[0] += 1 	# increase look ahead



	""" Events handling """
	def event(self, e):
		if e.type == pygame.KEYDOWN and self.currentMaxStep < self.LSCodeLen:
			if e.unicode == 'n':
				self.inc_max_step()
			if e.unicode == 'r':
				self.autorun = not self.autorun
			if e.unicode == '+':
				Conf.LSYSTEM.AUTORUN_STEP *= 10
			if e.unicode == '-':
				Conf.LSYSTEM.AUTORUN_STEP /= 10

	""" This function allow to manage the step by step turtle running """
	def update(self):
		if self.stepbystep:
			self.turtle.begin()
			if self.autorun:
				self.inc_max_step(step=Conf.LSYSTEM.AUTORUN_STEP)
			while self.currentStep < self.currentMaxStep:
				#pdb.set_trace()
				tmp = [self.currentStep];
				self.execute(tmp)
				self.currentStep = tmp[0]
			self.turtle.end()


	def generate_rec(self, iterations):
		if iterations == 0:
			return
		newcode = ''.join(map(self.get_rule, self.LSCode))


		self.LSCode = newcode

		print "============= Generation [" + str(self.LSSteps - iterations) + "], size=" + str(len(newcode)) + ": "
		if 'lsystem' in Conf.DEBUG and Conf.DEBUG['lsystem'] >= 2:
			print newcode

		self.generate_rec(iterations - 1)

	# Allow to recursively generate the code, keeping attention to the code parameters
	def generate_param_rec(self, itern):
		#pdb.set_trace()
		self.LSCodeLen = len(self.LSCode)
		if itern == 0:
			return

		rplc_list = []
		la = [0] # look ahead

		# traverse the code string and create the replacement string lists
		while la[0] < self.LSCodeLen:
			rplc_list.append(self.apply_rule(la))

		newcode = ''.join(rplc_list)

		self.LSCode = newcode

		print "============= Generation [" + str(self.LSSteps - itern + 1) + "], size=" + str(len(newcode)) + ": "
		if 'lsystem' in Conf.DEBUG and Conf.DEBUG['lsystem'] >= 2:
			print newcode

		self.generate_param_rec(itern - 1)

	# build the replacement string of the current token
	# move the look ahead
	def apply_rule(self, la):
		char = self.LSCode[la[0]] # get current character
		r = self.rules_contain(char) # get the corresponding rule
		#pdb.set_trace()
		if r: # if rule is set, we need to find the parameter given to the rule
			arg = self.find_arg(la)
			# None arg means that the rule is not parametized
			if not arg:
				return self.LSRules[r] # just return the rule
			param = self.find_param(r) # find the param name of the rule
			parametized_rule = self.LSRules[r].replace(param, arg) # replace the param name by its value
			#print char, '(', arg, ')', '=', parametized_rule
			#if parametized_rule == 'F(x/2)+F(x/2/2)':
			#	pdb.set_trace()
			return parametized_rule

		else: # if no rule is defined, just return the character
			la[0] += 1 	# increase look ahead
			return char

	def find_param(self, r):
		ob = 0
		cb = 0
		for k in range(len(r)):
			if r[k] == '(':
				ob = k
			if r[k] == ')':
				cb = k
		if cb is 0 or cb is ob:
			return None
		return r[ob+1:cb]


	# find the argument given to this function
	# la must point on the name of the function
	def find_arg(self, la):
		# go to next caracter
		la[0] += 1
		if la[0] >= self.LSCodeLen:
			return None
		# if current character is not '('
		# we are not on a parametized rule
		if self.LSCode[la[0]] != '(':
			return None
		k = la[0]
		while self.LSCode[k] != ')':
			k += 1 # find the closing bracket
		res = self.LSCode[la[0]+1:k] # substring
		la[0] = k + 1;
		return res


	def rules_contain(self, char):
		for r in self.LSRules:
			if r[0] == char: # the rule is 1char named
				return r
		return False

	def get_rule(self, char):
		if char in self.LSRules:
			return self.LSRules[char]
		else:
			return char

	def inc_max_step(self, step=1):
		if self.currentMaxStep < self.LSCodeLen:
			self.currentMaxStep += step
			while self.currentMaxStep < self.LSCodeLen and self.LSCode[self.currentMaxStep] not in self.LSVars:
				self.currentMaxStep += 1
		if self.currentMaxStep >= self.LSCodeLen:
			self.currentMaxStep = self.LSCodeLen - 1

File: src/js/LSystem.js

/**
 * Implement Mechanism for Lindenmayer System generation.
 * The rules and variables are specified in a concrete class
 * Inherit this class to implement a new L-System]
 */
class LSystem {
    /**
     *
     * @param {Turtle} turtle turtle instance for drawing the generated fractal
     * @param {Object} lsSpecs l-system specifications object. This should have the fields
     *  * misc: L-System misc details as an object with the fields:
     *      * name: name of the fractal
     *      * url: url to a wikipedia or other webpage containing fractal description
     *      * angle: default angle applied to rotations
     *      * segment: default segment length applied to forward or backward steps
     *      * steps: default number of steps (when not specified by the user)
     *      * dna: starting code
     *      * drawType: the way the fractal should be rendered
     *                  ('lines', 'line_strip', 'triangles', 'triangle_strip', etc...)
     * * vars: L-System variables, as an object with a field for all symbols specified
     *         in the fractal dna except rules. Values should be the name of the turtle's method.
     *         Symbols not specified as variables will be ignored when drawing.
     * * params: L-System variable parameters, distance / angle / change amount for all
     *           symbols specified in the variables set
     * * rules: L-System replacement rules, indicates what string should a symbol be replaced
     *          with during the fractal's iteration.
     * @param {Boolean} debug debug mode, will print generated fractal code
     */
    constructor(turtle, lsSpecs, debug, id) {
        this.turtle = turtle;
        const {
            misc,
            vars,
            params,
            rules
        } = lsSpecs;
        this.lsSpecs = lsSpecs;

        this.id = id;
        this.LSName = misc.name || 'Undefined (set lsSpecs.misc.name)';
        this.LSSteps = misc.steps || 7;
        this.LSDNA = misc.dna || 'F';
        this.LSDrawType = misc.drawType || 'line_strip';
        this.LSDrawType = this.LSDrawType.toLowerCase();

        this.LSRules = rules || {};
        this.LSVars = vars || {};
        this.LSParams = params || {};

        this.LSStochastic = false;
        this.LSStochRange = 0.01;

        this.LSVertexShader = "";
        this.LSPixelShader = "";

        this.currentStep = 0;
        this.currentMaxStep = 0;
        this.autorun = false;
        this.autorunStep = 5;
        this.stepByStep = false;
        this.LSCode = this.LSDNA;

        this.debug = debug;
        this._lastLog = 0;
        this._generationStart = 0;
        this._turtleRunStart = 0;
        this._interrupt = false;
        this._bindEvents();
    }

    /**
     * Interrupt whichever operation the lsystem is currently performing.
     * Useful if willing to interrupt an on-going async generation
     */
    interrupt() {
        this._interrupt = true;
    }

    /**
     * Run one step of the update loop. Called before every frame rendering.
     */
    update() {
        if (this.stepByStep) {

            this.turtle.begin();

            if (this.autorun) {
                this.incrementMaxStep(this.autorunStep);
            }
            this._executionLoopStepByStep();

            this.turtle.end();
        }
    }

    /**
     * Override default number of steps defined by the constructor.
     * @param {Integer|Float} n number of steps
     */
    setSteps(n) {
        this.LSSteps = n;
        if (this.LSSteps < 1) {
            this.LSSteps = 1;
        }
    }

    /**
     * Generate the code of the fractal from rules and DNA.
     * @param {Function} [done] - done callback. If provided, generation will be asynchroneous
     *      to avoid blocking the main thread. Done willbe called when generation is complete.
     */
    generate(done) {
        this._generationStart = Date.now();
        this._interrupt = false;
        this.LSCode = this.LSDNA;
        logging.log("[LSystem] Generation Started");
        logging.log("[LSystem] Fractal Name: " + this.LSName);
        logging.log("[LSystem] Precision: " + this.LSSteps);
        if (this.LSStochastic) {
            logging.log("[LSystem] Stochastic Factor: " + this.LSStochRange);
        }
        logging.log("[LSystem] Initial Axiom (DNA): " + this.LSDNA);
        logging.log("[LSystem] Rules: ");
        for (const rule in this.LSRules) {
            if (this.LSRules.hasOwnProperty(rule)) {
                const replacement = this.LSRules[rule];
                logging.log('    ' + rule + ' => ' + replacement);
            }
        }

        if (done) {
            this._runGenerationLoopsAsync(() => {
                const t = Date.now() - this._generationStart;
                logging.log("[LSystem] Async Generation Complete (l=" + this.LSCode.length + " t=" + t + "ms)!");
                if (!this._interrupt) { done(); };
            });
        }
        else {
            this._runGenerationLoops();
            const t = Date.now() - this._generationStart;
            logging.log("[LSystem] Generation Complete (l=" + this.LSCode.length + " t=" + t + "ms)!");
        }

    }

    /**
     * Run the turtle over the generated fractal to build the 3D fractal
     * @param {Boolean} stepByStep - step-by-step mode will draw the fractal as it is being generated
     * @param {Function} done - done callback. If provided the running of the turtle will be async
     *      This should avoid blocking the main thred when generating a large fractal.
     */
    runTurtleRun(stepByStep, done) {
        this._turtleRunStart = Date.now();
        this.turtle.reinit();

        this.stepByStep = stepByStep;
        this.turtle.setDrawType(this.LSDrawType);

        if (this.stepByStep) {
            this.currentStep = 0;
            this.currentMaxStep = 0;
            this.incrementMaxStep();
        }

        this.turtle.begin();

        const _done = () => {
            this.turtle.end();
            if (!this.stepByStep) {
                this._turtleLogCompleted(!!done);
            }
            if (done && !this._interrupt) {
                return done();
            }
        }

        if (this.stepByStep) {
            if (done) {
                return this._executionLoopStepByStep(_done);
            }
            this._executionLoopStepByStep();
            return _done();
        }
        if (done) {
            return this._executionLoop(_done);
        }
        this._executionLoop();
        return _done();
    }

    _turtleLog(count, force) {
        if (force || this.turtle.vertexBufferLen > 1 && new Date() - this._lastLog > 1000) {
            const pc = count * 100 / this.LSCode.length
            const _pc = Math.round(pc * 1000) / 1000;
            const t = Date.now() - this._turtleRunStart;
            logging.log("[LSYSTEM][" + t + "ms] > Turtle is running [" + _pc + '%], [' + this.turtle.vertexBufferLen + '] vertex');
            this._lastLog = new Date();
        }
    }

    _turtleLogCompleted(async) {
        this.stepByStep = false;
        const l = this.turtle.vertexBufferLen;
        const t = Date.now() - this._turtleRunStart;
        logging.log("[LSYSTEM] " + (async ? 'Async t' : 'T') + "urtle run completed (l=" + l + ", t=" + t + "ms)!")
    }

    _runGenerationLoops() {
        let rplc_list = [];
        let la = 0;
        let newCode = '';
        for (let step = 0; step < this.LSSteps; step++) {
            rplc_list = [];
            la = 0;

            while (la < this.LSCode.length) {
                const {
                    replacement,
                    laInc
                } = this._applyRule(la);
                rplc_list.push(replacement);
                la += laInc;
            }

            newCode = rplc_list.join('');

            this._generationLog(step, rplc_list);
            if (this.debug) {
                console.log(newCode);
            }
            this.LSCode = newCode;
        }
        this._generationLog(this.LSSteps, [], true);
    }

    _runGenerationLoopsAsync(done) {
        let step = 0
        let look_ahead = 0;
        let rplc_list = [];
        const BATCH_SIZE = 100000;

        const runOneStep = () => {
            rplc_list = [];
            look_ahead = 0;

            if (!this._interrupt) {
                return runBatchTokens();
            }
        }

        const runBatchTokens = () => {
            for (let index = 0; index < BATCH_SIZE; index++) {
                const {
                    replacement,
                    laInc
                } = this._applyRule(look_ahead);
                rplc_list.push(replacement);
                look_ahead += laInc;

                this._generationLog(step, rplc_list);

                if (look_ahead >= this.LSCode.length) { break; }
            }

            this._generationLog(step, rplc_list);

            if (look_ahead < this.LSCode.length) {
                if (!this._interrupt) {
                    return runBatchTokens();
                }
            }
            else {
                if (!this._interrupt) {
                    return completeStep();
                }
            }
        }

        const completeStep = () => {
            const newCode = rplc_list.join('');

            if (this.debug) {
                console.log(newCode);
            }
            this.LSCode = newCode;

            if (step < this.LSSteps) {
                step += 1;
                if (!this._interrupt) {
                    return setTimeout(runOneStep, 0)
                }
            }
            else {
                this._generationLog(step, [], true);
                if (done && !this._interrupt) {
                    done();
                }
            }
        }

        return runOneStep();
    }

    _generationLog(step, rplc_list, force) {
        if (force || Date.now() - this._lastLog > 1000) {
            this._lastLog = Date.now();
            const duration = Date.now() - this._generationStart;
            const len = rplc_list.reduce((acc, s) => acc + s.length, 0);
            const size = Math.max(this.LSCode.length, len);
            logging.log("[LSystem][" + duration + "ms] > Generation [" + step + "], size=" + size);
        }
    }


    _applyRule(la) {
        const char = this.LSCode[la];
        const r = this._getFullRuleKey(char);

        // if rule is set, we need to find the parameter given to the rule
        if (r) {
            const {arg, laInc} = this._findArg(la);
            if (!arg) {
                return {
                    replacement: this.LSRules[r],
                    laInc
                };
            }
            const param = this._findParam(r);
            return {
                replacement: this.LSRules[r].split(param).join(arg),
                laInc
            };
        }
        else {
            return {
                replacement: char,
                laInc: 1
            };
        }
    }

    _findParam(rule) {
        let opening_bracket = 0;
        let closing_bracket = 0;
        for (let k = 0; k < rule.length; k++) {
            if (rule[k] == '(') { opening_bracket = k; }
            if (rule[k] == ')') { closing_bracket = k; }
        }

        if (opening_bracket === 0 || closing_bracket === 0) {
            return null;
        }
        return rule.slice(opening_bracket, closing_bracket);
    }

    _findArg(la) {
        let laInc = 1;
        if (la + laInc > this.LSCode.length) {
            return {laInc, arg: null};
        }
        if (this.LSCode[la + laInc] != '(') {
            return {laInc, arg: null};
        }
        while (this.LSCode[la + laInc] != ')') {
            laInc += 1;
        }

        // la is on the first letter of f(...)
        // la + laInc points to the closing bracket
        const res = this.LSCode.slice(la + 2, la + laInc)

        return {
            arg: null,
            laInc: laInc + 1
        };
    }

    _getFullRuleKey(char) {
        for (const rule in this.LSRules) {
            if (this.LSRules.hasOwnProperty(rule)) {
                // rule name can only have 1 character
                if (rule[0] === char) {
                    return rule;
                };
            }
        }
        return null;
    }

    incrementMaxStep(by=1) {
        if (this.currentMaxStep < this.LSCode.length) {
            this.currentMaxStep += by;
            // increment to the next drawing action
            while (this.currentMaxStep < this.LSCode.length
                   && this.LSVars[this.LSCode[this.currentMaxStep]] != 'forward'
                   && this.LSVars[this.LSCode[this.currentMaxStep]] != 'backward'
                   && this.LSVars[this.LSCode[this.currentMaxStep]] != 'point') {
                this.currentMaxStep += 1;
            }
        }
        if (this.currentMaxStep >= this.LSCode.length) {
            this.currentMaxStep = this.LSCode.length;
        }
    }

    _executionLoopStepByStep(done) {
        while (this.currentStep < this.currentMaxStep) {
            this.currentStep = this._execute(this.currentStep);
            this._turtleLog(this.currentStep);
        }
        if (this.currentStep >= this.LSCode.length - 1) {
            this._turtleLog(this.currentStep, true);
            this._turtleLogCompleted();
        }
        // in step-by step mode, async execution isn't necessary
        if (done) {
            done();
        }
    }

    _executionLoop(done) {
        let count = 0
        const BATCH_SIZE = 100000;

        const _executeBatchAsync = () => {
            for (let index = 0; index < BATCH_SIZE; index++) {
                count = this._execute(count);
                this._turtleLog(count);
                if (count >= this.LSCode.length || this._interrupt) { break; }
            }
            if (count < this.LSCode.length && !this._interrupt) {
                return setTimeout(_executeBatchAsync, 0);
            }
            this._turtleLog(count, true);
            if (!this._interrupt) {
                return done();
            }
        }
        if (done) {
            if (!this._interrupt) {
                return _executeBatchAsync();
            }
        }
        else {
            while (count < this.LSCode.length) {
                count = this._execute(count);
                this._turtleLog(count);
            }
            this._turtleLog(count, true);
        }
    }

    _execute(position) {
        var debug = position;
        const char = this.LSCode[position];
        if (position + 1 < this.LSCode.length && this.LSCode[position + 1] == '(') {
            let {
                arg,
                laInc
            } = this._findArg(position);
            if (arg === null) {
                console.error("Error: No argument found for instruction `" + char + " in: ",
                              this.LSCode.slice(debug, position + 10));
            }

            // if we're executing a parametized rule, replace each variable
            // by its value in the params array
            for (const param of this.LSParams) {
                if (arg.indexOf(param) >= 0) {
                    arg = arg.split(param).join(this.LSParams[param])
                }
            }
            const res = eval(arg);

            if (this.LSVars[char]) {
                this.turtle[this.LSVars[char]](res);
            }

            return position + laInc;
        }
        else {
            if (this.LSVars[char]) {
                this.turtle[this.LSVars[char]](this.LSParams[char]);
            }
            return position + 1;
        }
    }

    _bindEvents() {
        window.addEventListener('keydown', (e) => {
            if (this.currentMaxStep >= this.LSCode.length) {
                return;
            }

            if (e.key === 'n') {
                this.incrementMaxStep();
            }
            if (e.key === 's') {
                this.incrementMaxStep();
            }
            if (e.key === 'r') {
                this.autorun = !this.autorun;
            }
            if (e.key === '+') {
                this.autorunStep *= 10;
            }
            else if (e.key === '-') {
                this.autorunStep /= 10;
            }
        })
    }
}